documentation/man/man3/Map.3
.TH "Map" 3 "Tue Mar 17 2015" "Minautorus" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Map \- \fBMap\fP generation and displaying\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Map\&.hpp>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBMap\fP ()"
.br
.ti -1c
.RI "\fB~Map\fP ()"
.br
.ti -1c
.RI "void \fBdisplay\fP (sf::RenderWindow *win)"
.br
.RI "\fIDisplay the map on a RenderWindow\&. \fP"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBinitTiles\fP ()"
.br
.RI "\fIInitialize the tiles of the map\&. \fP"
.ti -1c
.RI "void \fBcreateVertexTile\fP (size_t i, size_t j, size_t w, size_t h, sf::Color color)"
.br
.RI "\fICreate a vertex in the vertex array\&. \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBTile\fP * \fBm_tiles\fP [\fBNB_TILE_HEIGHT\fP][\fBNB_TILE_WIDTH\fP]"
.br
.ti -1c
.RI "sf::VertexArray \fBm_vertex_array\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
\fBMap\fP generation and displaying\&. 


.PP
\fBAuthor:\fP
.RS 4
bog 
.RE
.PP

.PP
Definition at line 28 of file Map\&.hpp\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Map::Map ()"

.PP
Definition at line 4 of file Map\&.cpp\&.
.SS "Map::~Map ()"

.PP
Definition at line 59 of file Map\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void Map::createVertexTile (size_t i, size_t j, size_t w, size_t h, sf::Color color)\fC [private]\fP"

.PP
Create a vertex in the vertex array\&. 
.PP
\fBParameters:\fP
.RS 4
\fIi\fP the coordinate of the vertex to create (line) 
.br
\fIj\fP the coordinate of the vertex to create (column) 
.br
\fIw\fP the width of the vertex to create 
.br
\fIh\fP the height of the vertex to create 
.br
\fIcolor\fP color of the vertex to create
.RE
.PP
\fBAuthor:\fP
.RS 4
bog 
.RE
.PP

.PP
Definition at line 35 of file Map\&.cpp\&.
.SS "void Map::display (sf::RenderWindow * win)"

.PP
Display the map on a RenderWindow\&. 
.PP
\fBAuthor:\fP
.RS 4
bog 
.RE
.PP

.PP
Definition at line 11 of file Map\&.cpp\&.
.SS "void Map::initTiles ()\fC [private]\fP"

.PP
Initialize the tiles of the map\&. 
.PP
\fBAuthor:\fP
.RS 4
bog 
.RE
.PP

.PP
Definition at line 17 of file Map\&.cpp\&.
.SH "Member Data Documentation"
.PP 
.SS "\fBTile\fP* Map::m_tiles[\fBNB_TILE_HEIGHT\fP][\fBNB_TILE_WIDTH\fP]\fC [private]\fP"

.PP
Definition at line 43 of file Map\&.hpp\&.
.SS "sf::VertexArray Map::m_vertex_array\fC [private]\fP"

.PP
Definition at line 44 of file Map\&.hpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Minautorus from the source code\&.


src/Map.cpp
#include <Map.hpp>
#include <Maze.hpp>
#include <cassert>

//TODO USING CHUNKS

Map::Map()
{
  m_vertex_array = sf::VertexArray(sf::Quads, NB_TILE_HEIGHT * NB_TILE_WIDTH * 4);
  
  initTiles();
  
}


void Map::initTiles()
{
  /* init Tiles */
  for(size_t i=0; i < NB_TILE_HEIGHT; i++)
    {
        for(size_t j=0; j < NB_TILE_WIDTH; j++)
	  {
	    m_tiles[i][j] = new Tile;
	    m_tiles[i][j]->i = i;
	    m_tiles[i][j]->j = j;
	    m_tiles[i][j]->obstacle = false;
	    m_tiles[i][j]->type = NONE;
	  }
    }

  /* place a maze */
  placeMaze(0, 0, NB_BLOC_WIDTH, NB_BLOC_HEIGHT);

  /* init vertex array */
  for(size_t i=0; i < NB_TILE_HEIGHT; i++)
    {
        for(size_t j=0; j < NB_TILE_WIDTH; j++)
	  {
	    createVertexTile( m_tiles[i][j] );
	  }
    }

}

void Map::placeMaze(size_t I, size_t J, size_t W, size_t H)
{
  MazeGenerator::Maze maze(H, W);
  maze.compute();
 
  for( size_t i = 0; i < H; i++ )
    {
      for( size_t j = 0; j < W; j++ )
	{
	  bool *b = maze.get(i,j)->border;
	 
	  /* entrance and exit */
	  if( i == H-1 && j == W-1 )
	    {
	      b[3] = false ;
	    }

	  if( i == 0 && j == 0 )
	    {
	      b[2] = false ;
	    }
	  
	  putBloc(I + i * (BLOC_SIZE+1), J + j * (BLOC_SIZE+1),
		  BLOC_SIZE, BLOC_SIZE, b);
	}
    }
}

void Map::putBloc(size_t I, size_t J, size_t W, size_t H, bool border[4])
{
  assert(border);
  assert( I >= 0 );
  assert( I < NB_TILE_HEIGHT );
  assert( J >= 0 );
  assert( J < NB_TILE_WIDTH );
  
  for(size_t i = 0; i < H+2; i++)
    {
      for(size_t j = 0; j < W+2; j++)
	{
	 
	  if( ( i == 0 && border[0] ) )
	    {
	      m_tiles[i + I][j + J]->type = WALL_UP;
	    }
	  else if( ( i == H+1 && border[1] ) )
	    {
	      m_tiles[i + I][j + J]->type = WALL_DOWN;
	    }
	  else if( ( j == 0 && border[2] ) )
	    {
	      m_tiles[i + I][j + J]->type = WALL_LEFT;
	    }
	  else if( ( j == W+1 && border[3] ) )
	    {
	      m_tiles[i + I][j + J]->type = WALL_RIGHT;
	    }
	  else
	    {
	      // walls can't be removed
	      if( !isAWall(i + I, j + J) )
		{
		  m_tiles[i + I][j + J]->type = GROUND;
		}
	      
	    }
	  
	}
    }
}

void Map::createVertexTile(Tile *tile)
{
  assert(tile);
  
  sf::Color color;

  switch( tile->type )
    {
    case GROUND:
      color = sf::Color::Green;
      break;

    case WALL_UP:
    case WALL_DOWN:
    case WALL_LEFT:
    case WALL_RIGHT:
      color = sf::Color::Black;
      break;
      
    default:
      color = sf::Color::Red;
      break;
    }
  
  createVertexTile(tile->i, tile->j, TILE_WIDTH, TILE_HEIGHT,color);
}

void Map::createVertexTile(size_t i, size_t j,size_t w,size_t h, sf::Color color)
{

  //real coordinate in the windows
  size_t I = i * TILE_HEIGHT;
  size_t J = j * TILE_WIDTH;

  //get a pointer on the vertex at the coord (i,j)
  sf::Vertex *tile = &m_vertex_array[ (i * NB_TILE_WIDTH + j) * 4 ];
   

  tile[0].position = sf::Vector2f(J,I);
  tile[1].position = sf::Vector2f(J + w, I);
  tile[2].position = sf::Vector2f(J + w, I + h);
  tile[3].position = sf::Vector2f(J, I + h);

  //TODO Texture
  tile[0].color = color;
  tile[1].color = color;
  tile[2].color = color;
  tile[3].color = color;
  
}

bool Map::isAWall(size_t i, size_t j)
{
  assert(m_tiles[i][j]);
  
  switch( m_tiles[i][j]->type )
    {
    case WALL_UP:
    case WALL_DOWN:
    case WALL_LEFT:
    case WALL_RIGHT:
      return true;
	    
    default:
      return false;
      break;
    }
}

bool Map::isInTheScreen(Tile const* t)
{
  assert(t);
  
  return (t->i >= 0 && t->i < NB_VISIBLE_TILE_HEIGHT
	  && t->j >= 0 && t->j < NB_VISIBLE_TILE_WIDTH);
}


void Map::display(sf::RenderWindow *win)
{
  win->draw(m_vertex_array);
}

Map::~Map()
{
  //delete Tiles
  for(size_t i=0; i < NB_TILE_HEIGHT; i++)
    {
      for(size_t j=0; j < NB_TILE_WIDTH; j++)
	{
	  delete m_tiles[i][j];

	}
    }

  m_vertex_array.clear();

}
